source/core/slang-com-ptr.h

```diff
@@ -51,21 +51,18 @@ struct Guid
 
 SLANG_FORCE_INLINE bool operator==(const Guid& aIn, const Guid& bIn)
 {
+	// Use the largest type the honors the alignment of Guid
+	typedef uint32_t CmpType;
     struct GuidCompare
     {
-        enum { kNum = sizeof(Guid) / sizeof(size_t) };
         Guid guid;
-        size_t data[kNum];
+		CmpType data[sizeof(Guid) / sizeof(CmpType)];
     };
-    const GuidCompare& a = reinterpret_cast<const GuidCompare&>(aIn);
-    const GuidCompare& b = reinterpret_cast<const GuidCompare&>(bIn);
-    
-    switch (GuidCompare::kNum)
-    {
-        case 2:     return ((a.data[0] ^ b.data[0]) | (a.data[1] ^ b.data[1])) == 0;   
-        case 4:     return ((a.data[0] ^ b.data[0]) | (a.data[1] ^ b.data[1]) | (a.data[2] ^ b.data[2]) | (a.data[3] ^ b.data[3]) ) == 0;   
-        default:    return false;
-    }
+	// Type pun - so compiler can 'see' the pun and not break aliasing rules 
```

Comment:
Strictly in C++ using a union to handle aliasing does not follow the letter of the law. 
In practice major compilers use the union as the mechanism to allow type punning - thus using a union is the 'de facto' method. The mechanism here may not be _strictly_ be following even this, because there is an argument based on an interpretation of the standards, that you should memcpy into the union and then access. But the standard also has a section about if you are not violating the 'layout' of the underlying types within a cast then you are okay. So from that point of view the first cast is okay - because the union contains a layout that exactly matches the original types layout - because it contains that type. From there it just accesses through the union - using the 'defacto' union mechanism.

https://stackoverflow.com/questions/25664848/unions-and-type-punning